File: python/problems_claude/2.py
# Problem 2: Move Zeros to End
# **Difficulty:** Easy
# **Time:** 15 minutes
# **Pattern:** Two-pointer technique (same direction)

# Move all zeros in an array to the end while maintaining the relative order of non-zero elements.

def move_zeros_to_end(arr):
    """
    Move all zeros to the end of the array while maintaining order of non-zero elements.

    Examples:
    Input: [0, 1, 0, 3, 12]
    Output: [1, 3, 12, 0, 0]

    Input: [0, 0, 1]
    Output: [1, 0, 0]
    """
    # Your code here
    n = len(arr)
    ret_array = [0] * n
    pos_pointer = 0
    for i in range(n):
        if arr[i] != 0:
            ret_array[pos_pointer] = arr[i]
            pos_pointer += 1

    print(f"ARRAY: {ret_array}")
    return ret_array


# Test cases
assert move_zeros_to_end([0, 1, 0, 3, 12]) == [1, 3, 12, 0, 0]
assert move_zeros_to_end([0, 0, 1]) == [1, 0, 0]
assert move_zeros_to_end([1, 2, 3]) == [1, 2, 3]
assert move_zeros_to_end([0]) == [0]
print("All tests passed!")

File: cpp/problems_1/2.cpp
// #### Problem 2: Valid Parentheses
// **Difficulty:** Easy
// **Time:** 15 minutes
// **Pattern:** Stack for matching pairs
//
// Given a string `s` containing just the characters `'('`, `')'`, `'{'`, `'}'`,
// `'['` and `']'`, determine if the input string is valid.

#include <cassert>
#include <iostream>
#include <map>
#include <print>
#include <stack>
#include <string>
#include <vector>
using namespace std;

bool isValid(std::string s) {
  /*
  Check if parentheses/brackets/braces are valid.

  Examples:
  Input: "()"
  Output: true

  Input: "()[]{}"
  Output: true

  Input: "(]"
  Output: false
  */

  // Your code here
  stack<char> my_stack;
  vector<char> left_chars = {'(', '{', '['};
  map<char, char> the_map = {{')', '('}, {'}', '{'}, {']', '['}};
  for (std::string::iterator it = s.begin(); it != s.end(); ++it) {
    bool found =
        find(left_chars.begin(), left_chars.end(), *it) != left_chars.end();
    if (found) {
      my_stack.push(*it);
    } else {
      char to_check = my_stack.top();
      my_stack.pop();
      if (the_map[*it] != to_check) {
        return false;
      }
    }
  }

  return true;
}

int main() {
  std::println("hello world");
  assert(isValid("()") == true);
  assert(isValid("()[]{}") == true);
  assert(isValid("(]") == false);
  assert(isValid("([)]") == false);
  assert(isValid("{[]}") == true);
  assert(isValid("") == true);

  std::cout << "Problem 2: All tests passed!" << std::endl;
  return 0;
}
